File: app/src/core/chladni.ts
/** Chladni — gerçek plaka fiziği (Kirchhoff–Love ince plaka teorisi).
 *
 * ÖNEMLİ FİZİKSEL GERÇEK: Bir Chladni deseni frekansın tek başına fonksiyonu DEĞİLDİR.
 * Aynı 528 Hz, farklı levhada tamamen farklı desen verir. Deseni belirleyen:
 *   malzeme (E, ρ, ν) · kalınlık h · kenar uzunluğu L · sınır koşulu · uyarım noktası
 *
 * Bu yüzden burada "frekans → desen" değil, "levha + frekans → rezonans modu" hesaplanır:
 *   1. Levha parametrelerinden plaka sertliği D ve modal frekanslar f(m,n) bulunur
 *   2. Verilen frekansa en yakın mod(lar) seçilir
 *   3. Rezonansa yakınlığa göre modlar ağırlıklı toplanır (gerçek levha da böyle davranır)
 *   4. Düğüm çizgileri (|w| ≈ 0) kum birikim bölgeleridir
 *
 * Kaynak denklem (serbest kenarlı kare plaka, Ritz yaklaşımı):
 *   f(m,n) = (λ(m,n) / 2π) · √(D / (ρ·h·L⁴)),   D = E·h³ / (12(1−ν²))
 */
import type { PatternFingerprint } from "../data/types";
import { vectorSimilarity } from "./matching";

/** Malzeme sabitleri: E = Young modülü (Pa), rho = yoğunluk (kg/m³), nu = Poisson oranı. */
export interface Material {
  E: number;
  rho: number;
  nu: number;
}

export const MATERIALS: Record<string, Material> = {
  steel: { E: 200e9, rho: 7850, nu: 0.30 },
  aluminum: { E: 69e9, rho: 2700, nu: 0.33 },
  brass: { E: 100e9, rho: 8500, nu: 0.34 },
  glass: { E: 70e9, rho: 2500, nu: 0.22 },
};

export type BoundaryCondition = "free" | "clamped" | "simply_supported";

export interface Plate {
  material: keyof typeof MATERIALS | string;
  /** Kenar uzunluğu (cm) */
  sizeCm: number;
  /** Kalınlık (mm) */
  thicknessMm: number;
  boundary: BoundaryCondition;
}

export const DEFAULT_PLATE: Plate = { material: "steel", sizeCm: 20, thicknessMm: 1, boundary: "free" };

/** Plaka eğilme sertliği D = E·h³ / (12(1−ν²)) */
export const flexuralRigidity = (mat: Material, thicknessM: number): number =>
  (mat.E * thicknessM ** 3) / (12 * (1 - mat.nu ** 2));

/**
 * Boyutsuz frekans parametresi λ(m,n). Sınır koşuluna göre değişir.
 *
 * - simply_supported: analitik tam çözüm  λ = π²(m² + n²)
 * - clamped / free: Ritz yaklaşımı — kenar koşulu etkin dalga sayısını kaydırır.
 *   (Serbest kenar için Leissa'nın kare plaka tablolarına yaklaşan ampirik düzeltme.)
 */
export const lambdaMN = (m: number, n: number, boundary: BoundaryCondition): number => {
  if (boundary === "simply_supported") return Math.PI ** 2 * (m * m + n * n);
  // Ankastre ve serbest kenarda etkin dalga sayısı yarım dalga kadar kayar.
  const shift = boundary === "clamped" ? 0.5 : -0.5;
  const me = Math.max(0.5, m + shift);
  const ne = Math.max(0.5, n + shift);
  return Math.PI ** 2 * (me * me + ne * ne);
};

/** (m,n) modunun rezonans frekansı (Hz). */
export const modalFrequency = (m: number, n: number, plate: Plate): number => {
  const mat = MATERIALS[plate.material] ?? MATERIALS.steel;
  const h = plate.thicknessMm / 1000;
  const L = plate.sizeCm / 100;
  const D = flexuralRigidity(mat, h);
  return (lambdaMN(m, n, plate.boundary) / (2 * Math.PI)) * Math.sqrt(D / (mat.rho * h * L ** 4));
};

export interface Mode {
  m: number;
  n: number;
  hz: number;
  /** Sürülen frekansa göreli genlik (rezonans yakınlığı) */
  weight: number;
}

/** Sönüm oranı — gerçek metal levhalarda tipik olarak %0.1–1. */
const DAMPING = 0.006;

/**
 * Sürülen frekansta uyarılan modlar. Gerçek bir levha tek modda değil, rezonansa
 * yakınlıkla ağırlıklanmış mod süperpozisyonunda titreşir (Lorentz tepkisi):
 *   A(f) = 1 / √[(1 − (f/fᵢ)²)² + (2ζ·f/fᵢ)²]
 */
export const excitedModes = (hz: number, plate: Plate, maxOrder = 12, keep = 6): Mode[] => {
  if (hz <= 0) return [];
  const modes: Mode[] = [];
  for (let m = 1; m <= maxOrder; m++)
    for (let n = m; n <= maxOrder; n++) {
      const fi = modalFrequency(m, n, plate);
      if (!Number.isFinite(fi) || fi <= 0) continue;
      const r = hz / fi;
      const weight = 1 / Math.sqrt((1 - r * r) ** 2 + (2 * DAMPING * r) ** 2);
      modes.push({ m, n, hz: fi, weight });
    }
  const top = modes.sort((a, b) => b.weight - a.weight).slice(0, keep);
  const max = top[0]?.weight || 1;
  return top.map((x) => ({ ...x, weight: x.weight / max }));
};

/** Sürülen frekansa en yakın tek mod (özet gösterim için). */
export const dominantMode = (hz: number, plate: Plate = DEFAULT_PLATE): Mode | null => excitedModes(hz, plate, 12, 1)[0] ?? null;

/** Bir modun şekil fonksiyonu (sınır koşuluna göre). */
const shape = (m: number, n: number, x: number, y: number, boundary: BoundaryCondition): number => {
  if (boundary === "simply_supported") {
    return Math.sin(m * Math.PI * x) * Math.sin(n * Math.PI * y);
  }
  if (boundary === "clamped") {
    // Ankastre kenar: kenarlarda hem yer değiştirme hem eğim sıfır
    const f = (k: number, t: number) => (1 - Math.cos(2 * Math.PI * k * t)) / 2;
    return f(m, x) * f(n, y) * Math.sin(m * Math.PI * x) * Math.sin(n * Math.PI * y);
  }
  // Serbest kenar (klasik Chladni levhası): kosinüs tabanı, simetrik/antisimetrik birleşim
  return (
    Math.cos(m * Math.PI * x) * Math.cos(n * Math.PI * y) -
    Math.cos(n * Math.PI * x) * Math.cos(m * Math.PI * y)
  );
};

/** Sürülen frekansta levha yüzeyinin yer değiştirme alanı (süperpozisyon). */
export const displacementField = (hz: number, plate: Plate, size: number): Float32Array => {
  const modes = excitedModes(hz, plate);
  const out = new Float32Array(size * size);
  if (!modes.length) return out;
  for (let j = 0; j < size; j++) {
    const y = j / (size - 1);
    for (let i = 0; i < size; i++) {
      const x = i / (size - 1);
      let v = 0;
      for (const md of modes) v += md.weight * shape(md.m, md.n, x, y, plate.boundary);
      out[j * size + i] = v;
    }
  }
  let max = 0;
  for (let i = 0; i < out.length; i++) max = Math.max(max, Math.abs(out[i]));
  if (max > 0) for (let i = 0; i < out.length; i++) out[i] /= max;
  return out;
};

/**
 * Kum maskesi: kum, yer değiştirmenin sıfıra yakın olduğu düğüm çizgilerinde birikir.
 * Eşik levha genliğine bağlıdır — güçlü uyarımda çizgiler incelir.
 */
export const simulatePlate = (hz: number, plate: Plate = DEFAULT_PLATE, size = 128, threshold = 0.06): Uint8Array => {
  const field = displacementField(hz, plate, size);
  const mask = new Uint8Array(size * size);
  for (let i = 0; i < field.length; i++) mask[i] = Math.abs(field[i]) < threshold ? 1 : 0;
  return mask;
};

/** Geriye dönük uyum: varsayılan levha ile eski çağrı biçimi. */
export const simulate = (hz: number, size = 128, threshold = 0.06): Uint8Array =>
  simulatePlate(hz, DEFAULT_PLATE, size, threshold);

/** Eski API — artık gerçek modal hesaptan gelir. */
export const modeNumbers = (hz: number, plate: Plate = DEFAULT_PLATE): [number, number] => {
  const d = dominantMode(hz, plate);
  return d ? [d.m, d.n] : [1, 2];
};

/** Bu levhanın rezonans frekansları — desenin en net göründüğü noktalar. */
export const resonances = (plate: Plate = DEFAULT_PLATE, lo = 20, hi = 20000, maxOrder = 12): Mode[] => {
  const out: Mode[] = [];
  for (let m = 1; m <= maxOrder; m++)
    for (let n = m; n <= maxOrder; n++) {
      const hz = modalFrequency(m, n, plate);
      if (hz >= lo && hz <= hi) out.push({ m, n, hz, weight: 1 });
    }
  return out.sort((a, b) => a.hz - b.hz);
};

/** Sürülen frekans bir rezonansa ne kadar yakın? 1 = tam rezonans, 0 = uzak. */
export const resonanceProximity = (hz: number, plate: Plate = DEFAULT_PLATE): number => {
  const d = dominantMode(hz, plate);
  if (!d) return 0;
  const cents = Math.abs(1200 * Math.log2(hz / d.hz));
  return Math.max(0, 1 - cents / 200);         // 200 cent uzakta sıfıra iner
};

// ---------- PARMAK İZİ ----------

const mirrorScore = (a: Uint8Array, s: number, map: (i: number, j: number) => [number, number]): number => {
  let diff = 0;
  for (let j = 0; j < s; j++)
    for (let i = 0; i < s; i++) {
      const [i2, j2] = map(i, j);
      diff += Math.abs(a[j * s + i] - a[j2 * s + i2]);
    }
  return 1 - diff / (s * s);
};

export const fingerprint = (mask: Uint8Array, size: number): PatternFingerprint => {
  const s = size;
  const symmetry = (mirrorScore(mask, s, (i, j) => [s - 1 - i, j]) + mirrorScore(mask, s, (i, j) => [i, s - 1 - j]) + mirrorScore(mask, s, (i, j) => [j, i])) / 3;
  const radialSymmetry = (mirrorScore(mask, s, (i, j) => [s - 1 - j, i]) + mirrorScore(mask, s, (i, j) => [s - 1 - i, s - 1 - j])) / 2;
  let ones = 0, edges = 0;
  for (let j = 0; j < s; j++)
    for (let i = 0; i < s; i++) {
      const v = mask[j * s + i];
      ones += v;
      if (i + 1 < s) edges += Math.abs(v - mask[j * s + i + 1]);
      if (j + 1 < s) edges += Math.abs(v - mask[(j + 1) * s + i]);
    }
  let nodes = 0;
  const mid = Math.floor(s / 2);
  for (let i = 1; i < s; i++) {
    if (mask[mid * s + i] - mask[mid * s + i - 1] === 1) nodes++;
    if (mask[i * s + mid] - mask[(i - 1) * s + mid] === 1) nodes++;
  }
  const hist = new Array<number>(18).fill(0);
  let magSum = 0;
  for (let j = 1; j < s - 1; j++)
    for (let i = 1; i < s - 1; i++) {
      const gx = (mask[j * s + i + 1] - mask[j * s + i - 1]) / 2;
      const gy = (mask[(j + 1) * s + i] - mask[(j - 1) * s + i]) / 2;
      const mag = Math.hypot(gx, gy);
      if (mag === 0) continue;
      hist[Math.min(17, Math.floor((((Math.atan2(gy, gx) * 180) / Math.PI + 180) % 180) / 10))] += mag;
      magSum += mag;
    }
  return {
    symmetry,
    radialSymmetry,
    complexity: Math.min(1, edges / (2 * s * s * 0.25)),
    density: ones / (s * s),
    nodeCount: nodes,
    dominantAngleDeg: magSum > 0 ? hist.indexOf(Math.max(...hist)) * 10 + 5 : 0,
  };
};

export const fingerprintVector = (f: PatternFingerprint): number[] => [
  f.symmetry, f.radialSymmetry, f.complexity, f.density, Math.min(f.nodeCount, 64) / 64, f.dominantAngleDeg / 180,
];

export const similarity = (a: PatternFingerprint, b: PatternFingerprint): number =>
  vectorSimilarity(fingerprintVector(a), fingerprintVector(b));

/** Gri (0–255) → Otsu eşiği ile kum maskesi. */
export const grayToMask = (gray: Uint8ClampedArray | number[], size: number): Uint8Array => {
  const hist = new Array<number>(256).fill(0);
  for (let i = 0; i < size * size; i++) hist[gray[i] | 0]++;
  const total = size * size;
  let sumAll = 0;
  for (let i = 0; i < 256; i++) sumAll += i * hist[i];
  let w0 = 0, sum0 = 0, best = 128, bestVar = -1;
  for (let t = 0; t < 256; t++) {
    w0 += hist[t];
    if (w0 === 0 || w0 === total) continue;
    sum0 += t * hist[t];
    const m0 = sum0 / w0, m1 = (sumAll - sum0) / (total - w0);
    const v = w0 * (total - w0) * (m0 - m1) ** 2;
    if (v > bestVar) { bestVar = v; best = t; }
  }
  const out = new Uint8Array(size * size);
  for (let i = 0; i < size * size; i++) out[i] = gray[i] >= best ? 1 : 0;
  return out;
};

/** Rezonans haritası — bu levhanın gerçek modal frekansları işaretlenir. */
export const resonanceMap = (lo: number, hi: number, step: number, plate: Plate = DEFAULT_PLATE, size = 48) => {
  const rows: Array<{ hz: number; mode: [number, number]; complexity: number; nodeCount: number; proximity: number }> = [];
  for (let f = lo; f <= hi; f += step) {
    const fp = fingerprint(simulatePlate(f, plate, size), size);
    rows.push({ hz: f, mode: modeNumbers(f, plate), complexity: fp.complexity, nodeCount: fp.nodeCount, proximity: resonanceProximity(f, plate) });
  }
  return rows;
};

File: engine/aurora/chladni.py
"""Chladni Engine — gerçek plaka fiziği (Kirchhoff–Love) + görüntü parmak izi.

Simülasyon artık kaba bir yaklaşım değil, levha parametrelerinden (malzeme, kalınlık,
boyut, sınır koşulu) türetilen modal analizdir. Yine de 'simulation' etiketi taşır:
gerçek bir levhada kenar tutuşu, uyarım noktası ve malzeme homojensizliği deseni değiştirir.
""" 
from __future__ import annotations

from dataclasses import asdict

import numpy as np

from .models import PatternFingerprint, PatternRecord


# ---------- PLAKA FİZİĞİ (Kirchhoff–Love) ----------
#
# ÖNEMLİ: Chladni deseni frekansın tek başına fonksiyonu DEĞİLDİR. Aynı 528 Hz farklı
# levhada tamamen farklı desen verir. Deseni belirleyen: malzeme (E, ρ, ν), kalınlık h,
# kenar uzunluğu L, sınır koşulu ve uyarım noktası.
#
#   D = E·h³ / (12(1−ν²))                      plaka eğilme sertliği
#   f(m,n) = λ(m,n)/(2π) · √(D / (ρ·h·L⁴))     modal frekans

MATERIALS: dict[str, dict[str, float]] = {
    "steel":    {"E": 200e9, "rho": 7850, "nu": 0.30},
    "aluminum": {"E":  69e9, "rho": 2700, "nu": 0.33},
    "brass":    {"E": 100e9, "rho": 8500, "nu": 0.34},
    "glass":    {"E":  70e9, "rho": 2500, "nu": 0.22},
}

DEFAULT_PLATE = {"material": "steel", "size_cm": 20.0, "thickness_mm": 1.0, "boundary": "free"}
DAMPING = 0.006          # gerçek metal levhalarda tipik sönüm oranı


def flexural_rigidity(material: str, thickness_m: float) -> float:
    m = MATERIALS.get(material, MATERIALS["steel"])
    return m["E"] * thickness_m ** 3 / (12 * (1 - m["nu"] ** 2))


def lambda_mn(m: int, n: int, boundary: str) -> float:
    """Boyutsuz frekans parametresi. simply_supported analitik; diğerleri Ritz yaklaşımı."""
    if boundary == "simply_supported":
        return np.pi ** 2 * (m * m + n * n)
    shift = 0.5 if boundary == "clamped" else -0.5
    me, ne = max(0.5, m + shift), max(0.5, n + shift)
    return np.pi ** 2 * (me * me + ne * ne)


def modal_frequency(m: int, n: int, plate: dict | None = None) -> float:
    p = {**DEFAULT_PLATE, **(plate or {})}
    mat = MATERIALS.get(p["material"], MATERIALS["steel"])
    h, L = p["thickness_mm"] / 1000.0, p["size_cm"] / 100.0
    D = flexural_rigidity(p["material"], h)
    return float(lambda_mn(m, n, p["boundary"]) / (2 * np.pi) * np.sqrt(D / (mat["rho"] * h * L ** 4)))


def excited_modes(freq_hz: float, plate: dict | None = None, max_order: int = 12,
                  keep: int = 6) -> list[dict]:
    """Sürülen frekansta uyarılan modlar (Lorentz rezonans tepkisi ile ağırlıklı)."""
    if freq_hz <= 0:
        return []
    out = []
    for m in range(1, max_order + 1):
        for n in range(m, max_order + 1):
            fi = modal_frequency(m, n, plate)
            if not np.isfinite(fi) or fi <= 0:
                continue
            r = freq_hz / fi
            w = 1.0 / np.sqrt((1 - r * r) ** 2 + (2 * DAMPING * r) ** 2)
            out.append({"m": m, "n": n, "hz": fi, "weight": float(w)})
    top = sorted(out, key=lambda x: -x["weight"])[:keep]
    mx = top[0]["weight"] if top else 1.0
    return [{**x, "weight": x["weight"] / mx} for x in top]


def mode_numbers(freq_hz: float, plate: dict | None = None) -> tuple[int, int]:
    """Sürülen frekansa en yakın rezonans modu."""
    modes = excited_modes(freq_hz, plate, keep=1)
    return (modes[0]["m"], modes[0]["n"]) if modes else (1, 2)


def _shape(m: int, n: int, X: np.ndarray, Y: np.ndarray, boundary: str) -> np.ndarray:
    if boundary == "simply_supported":
        return np.sin(m * np.pi * X) * np.sin(n * np.pi * Y)
    if boundary == "clamped":
        f = lambda k, t: (1 - np.cos(2 * np.pi * k * t)) / 2
        return f(m, X) * f(n, Y) * np.sin(m * np.pi * X) * np.sin(n * np.pi * Y)
    return (np.cos(m * np.pi * X) * np.cos(n * np.pi * Y)
            - np.cos(n * np.pi * X) * np.cos(m * np.pi * Y))


def displacement_field(freq_hz: float, plate: dict | None = None, size: int = 128) -> np.ndarray:
    """Levha yüzeyinin yer değiştirme alanı (uyarılan modların süperpozisyonu)."""
    p = {**DEFAULT_PLATE, **(plate or {})}
    modes = excited_modes(freq_hz, p)
    x = np.linspace(0, 1, size)
    X, Y = np.meshgrid(x, x)
    field = np.zeros((size, size))
    for md in modes:
        field += md["weight"] * _shape(md["m"], md["n"], X, Y, p["boundary"])
    mx = np.abs(field).max()
    return field / mx if mx > 0 else field


def simulate(freq_hz: float, size: int = 128, threshold: float = 0.06,
             plate: dict | None = None) -> np.ndarray:
    """Kum maskesi: kum, yer değiştirmenin sıfıra yakın olduğu düğüm çizgilerinde birikir."""
    return (np.abs(displacement_field(freq_hz, plate, size)) < threshold).astype(np.uint8)


def resonances(plate: dict | None = None, lo: float = 20, hi: float = 20000,
               max_order: int = 12) -> list[dict]:
    """Bu levhanın gerçek rezonans frekansları — desenin en net göründüğü noktalar."""
    out = [{"m": m, "n": n, "hz": modal_frequency(m, n, plate)}
           for m in range(1, max_order + 1) for n in range(m, max_order + 1)]
    return sorted([x for x in out if lo <= x["hz"] <= hi], key=lambda x: x["hz"])


def resonance_proximity(freq_hz: float, plate: dict | None = None) -> float:
    """Sürülen frekans bir rezonansa ne kadar yakın? 1 = tam rezonans, 0 = uzak."""
    modes = excited_modes(freq_hz, plate, keep=1)
    if not modes:
        return 0.0
    c = abs(1200 * np.log2(freq_hz / modes[0]["hz"]))
    return float(max(0.0, 1 - c / 200))


def plate_field(n: int, m: int, size: int = 128, L: float = 1.0) -> np.ndarray:
    """Tek modun şekil fonksiyonu (geriye dönük uyum; ikon üretimi bunu kullanır)."""
    x = np.linspace(0, L, size)
    X, Y = np.meshgrid(x, x)
    return _shape(n, m, X / L, Y / L, "free")


# ---------- FINGERPRINT ----------

def _mirror_score(a: np.ndarray, b: np.ndarray) -> float:
    return float(1.0 - np.mean(np.abs(a.astype(float) - b.astype(float))))


def fingerprint(mask: np.ndarray) -> PatternFingerprint:
    """İkili maske (1 = kum) → sayısal parmak izi."""
    m = (np.asarray(mask) > 0).astype(np.uint8)
    if m.ndim != 2 or m.size == 0:
        raise ValueError("mask must be a non-empty 2D array")
    h, w = m.shape
    s = min(h, w)
    m = m[:s, :s]
    sym_lr = _mirror_score(m, m[:, ::-1])
    sym_ud = _mirror_score(m, m[::-1, :])
    sym_diag = _mirror_score(m, m.T)
    symmetry = (sym_lr + sym_ud + sym_diag) / 3.0
    radial = (_mirror_score(m, np.rot90(m, 1)) + _mirror_score(m, np.rot90(m, 2))) / 2.0
    density = float(m.mean())
    # karmaşıklık: kenar yoğunluğu (komşu farkı)
    edges = np.abs(np.diff(m.astype(int), axis=0)).sum() + np.abs(np.diff(m.astype(int), axis=1)).sum()
    complexity = float(min(1.0, edges / (2.0 * s * s * 0.25)))
    # düğüm sayısı: satır ortasındaki ve sütun ortasındaki geçişler
    mid_r, mid_c = m[s // 2, :], m[:, s // 2]
    node_count = int(np.sum(np.diff(mid_r.astype(int)) == 1) + np.sum(np.diff(mid_c.astype(int)) == 1))
    # baskın açı: gradient yönlerinin ağırlıklı histogramı
    gy, gx = np.gradient(m.astype(float))
    mag = np.hypot(gx, gy)
    ang = (np.degrees(np.arctan2(gy, gx)) % 180.0)
    hist, bins = np.histogram(ang, bins=18, range=(0, 180), weights=mag)
    dominant = float(bins[int(np.argmax(hist))] + 5.0) if mag.sum() > 0 else 0.0
    return PatternFingerprint(round(symmetry, 4), round(radial, 4), round(complexity, 4),
                              round(density, 4), node_count, round(dominant, 1))


def similarity(a: PatternFingerprint, b: PatternFingerprint) -> float:
    """0–1 arası görüntü/matematiksel desen benzerliği (enerji benzerliği DEĞİL)."""
    va, vb = np.asarray(a.as_vector()), np.asarray(b.as_vector())
    return float(max(0.0, 1.0 - np.linalg.norm(va - vb) / np.sqrt(len(va))))


def mask_similarity(a: np.ndarray, b: np.ndarray) -> float:
    """Piksel bazlı IoU (aynı boyuta yeniden örneklenir)."""
    a, b = (np.asarray(a) > 0), (np.asarray(b) > 0)
    s = min(a.shape[0], b.shape[0])
    ia = np.linspace(0, a.shape[0] - 1, s).astype(int)
    ib = np.linspace(0, b.shape[0] - 1, s).astype(int)
    a2, b2 = a[np.ix_(ia, ia)], b[np.ix_(ib, ib)]
    union = np.logical_or(a2, b2).sum()
    return float(np.logical_and(a2, b2).sum() / union) if union else 1.0


# ---------- REAL IMAGE ----------

def image_to_mask(gray: np.ndarray, threshold: float | None = None) -> np.ndarray:
    """Gri görüntü (0–255 veya 0–1) → kum maskesi. Kum açık renkli varsayılır (Otsu benzeri eşik)."""
    g = np.asarray(gray, dtype=float)
    if g.max() > 1.0:
        g = g / 255.0
    if threshold is None:
        # basit Otsu
        hist, edges = np.histogram(g, bins=64, range=(0, 1))
        total = g.size
        best_t, best_var = 0.5, -1.0
        w0 = 0.0
        sum_all = float((hist * ((edges[:-1] + edges[1:]) / 2)).sum())
        sum0 = 0.0
        for i, c in enumerate(hist):
            w0 += c
            if w0 == 0 or w0 == total:
                continue
            sum0 += c * ((edges[i] + edges[i + 1]) / 2)
            m0, m1 = sum0 / w0, (sum_all - sum0) / (total - w0)
            var = w0 * (total - w0) * (m0 - m1) ** 2
            if var > best_var:
                best_var, best_t = var, edges[i + 1]
        threshold = best_t
    return (g >= threshold).astype(np.uint8)


def load_image_gray(path: str) -> np.ndarray:
    try:
        from PIL import Image
    except ImportError as e:  # pragma: no cover
        raise RuntimeError("pip install pillow — görüntü analizi için gerekli") from e
    return np.asarray(Image.open(path).convert("L"), dtype=float)


def record_from_mask(pattern_id: str, hz: float, mask: np.ndarray, kind: str,
                     plate: dict | None = None, image_ref: str | None = None) -> PatternRecord:
    return PatternRecord(pattern_id, hz, kind, fingerprint(mask), plate or {}, image_ref)  # type: ignore[arg-type]


def resonance_map(lo: float, hi: float, step: float = 1.0, size: int = 64) -> list[dict]:
    """SIMULATION taraması: frekans → complexity/node_count. Rezonans bölgelerini görselleştirmek için."""
    out = []
    f = lo
    while f <= hi:
        fp = fingerprint(simulate(f, size))
        n, m = mode_numbers(f)
        out.append({"hz": round(f, 3), "mode": [n, m], "complexity": fp.complexity,
                    "node_count": fp.node_count, "kind": "simulation"})
        f += step
    return out


def record_to_dict(r: PatternRecord) -> dict:
    return {"pattern_id": r.pattern_id, "hz": r.hz, "kind": r.kind, "plate": r.plate,
            "image_ref": r.image_ref, "fingerprint": asdict(r.fingerprint)}
